File: go/jsonfmt.go
package entryconf

import (
	"math"
	"sort"
	"strconv"
	"strings"
)

// formatJSON writes a JSON-shaped value in the normalized form of SPEC §10.8:
// two-space indentation, one member or element per line, keys sorted by code
// point, minimal string escaping, integers as integers, one trailing newline.
func formatJSON(v any) string {
	var b strings.Builder
	writeJSON(&b, v, 0)
	b.WriteByte('\n')
	return b.String()
}

func writeJSON(b *strings.Builder, v any, depth int) {
	switch t := v.(type) {
	case nil:
		b.WriteString("null")
	case bool:
		if t {
			b.WriteString("true")
		} else {
			b.WriteString("false")
		}
	case string:
		writeJSONString(b, t)
	case int64:
		b.WriteString(strconv.FormatInt(t, 10))
	case float64:
		writeJSONNumber(b, t)
	case map[string]any:
		if len(t) == 0 {
			b.WriteString("{}")
			return
		}
		keys := make([]string, 0, len(t))
		for k := range t {
			keys = append(keys, k)
		}
		sort.Strings(keys) // Go compares strings bytewise = by code point for UTF-8
		b.WriteString("{\n")
		for i, k := range keys {
			indent(b, depth+1)
			writeJSONString(b, k)
			b.WriteString(": ")
			writeJSON(b, t[k], depth+1)
			if i+1 < len(keys) {
				b.WriteByte(',')
			}
			b.WriteByte('\n')
		}
		indent(b, depth)
		b.WriteByte('}')
	case []any:
		if len(t) == 0 {
			b.WriteString("[]")
			return
		}
		b.WriteString("[\n")
		for i, item := range t {
			indent(b, depth+1)
			writeJSON(b, item, depth+1)
			if i+1 < len(t) {
				b.WriteByte(',')
			}
			b.WriteByte('\n')
		}
		indent(b, depth)
		b.WriteByte(']')
	default:
		// Unreachable: values are normalized to the shapes above before use.
		b.WriteString("null")
	}
}

func indent(b *strings.Builder, depth int) {
	for i := 0; i < depth; i++ {
		b.WriteString("  ")
	}
}

// writeJSONNumber renders a float: integral values below 2^53 as plain
// integers, everything else in the shortest round-trip form (SPEC §10.8).
func writeJSONNumber(b *strings.Builder, f float64) {
	if f == math.Trunc(f) && math.Abs(f) < 1<<53 {
		b.WriteString(strconv.FormatInt(int64(f), 10))
		return
	}
	b.WriteString(strconv.FormatFloat(f, 'g', -1, 64))
}

func writeJSONString(b *strings.Builder, s string) {
	const hex = "0123456789abcdef"
	b.WriteByte('"')
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch c {
		case '"':
			b.WriteString(`\"`)
		case '\\':
			b.WriteString(`\\`)
		case '\b':
			b.WriteString(`\b`)
		case '\f':
			b.WriteString(`\f`)
		case '\n':
			b.WriteString(`\n`)
		case '\r':
			b.WriteString(`\r`)
		case '\t':
			b.WriteString(`\t`)
		default:
			if c < 0x20 {
				b.WriteString(`\u00`)
				b.WriteByte(hex[c>>4])
				b.WriteByte(hex[c&0xf])
			} else {
				b.WriteByte(c) // UTF-8 bytes pass through literally
			}
		}
	}
	b.WriteByte('"')
}

File: ts/src/jsonfmt.ts
import { isRecord } from "./pointer.ts";
import type { Value } from "./tree.ts";

/** Code-point order — UTF-8 byte order — not UTF-16 code-unit order. */
export function compareCodePoints(a: string, b: string): number {
  return Buffer.compare(Buffer.from(a, "utf8"), Buffer.from(b, "utf8"));
}

/**
 * Normalized JSON (SPEC §10.8): two-space indent, one member or element per
 * line, keys sorted by code point, minimal escaping, integers as integers, one
 * trailing newline.
 */
export function formatJSON(value: Value): string {
  return `${write(value, 0)}\n`;
}

function write(v: Value, depth: number): string {
  if (v === null) return "null";
  switch (typeof v) {
    case "boolean":
      return v ? "true" : "false";
    case "string":
      return quote(v);
    case "number":
      return Number.isInteger(v) && Math.abs(v) < 2 ** 53 ? String(v) : String(v);
  }
  const pad = "  ".repeat(depth + 1);
  const close = "  ".repeat(depth);
  if (Array.isArray(v)) {
    if (v.length === 0) return "[]";
    return `[\n${v.map((item) => pad + write(item, depth + 1)).join(",\n")}\n${close}]`;
  }
  if (isRecord(v)) {
    const keys = Object.keys(v).sort(compareCodePoints);
    if (keys.length === 0) return "{}";
    return `{\n${keys
      .map((k) => `${pad}${quote(k)}: ${write(v[k], depth + 1)}`)
      .join(",\n")}\n${close}}`;
  }
  return "null";
}

function quote(s: string): string {
  let out = '"';
  for (const c of s) {
    switch (c) {
      case '"':
        out += '\\"';
        break;
      case "\\":
        out += "\\\\";
        break;
      case "\b":
        out += "\\b";
        break;
      case "\f":
        out += "\\f";
        break;
      case "\n":
        out += "\\n";
        break;
      case "\r":
        out += "\\r";
        break;
      case "\t":
        out += "\\t";
        break;
      default: {
        const code = c.codePointAt(0)!;
        if (code < 0x20) out += `\\u00${code.toString(16).padStart(2, "0")}`;
        else out += c;
      }
    }
  }
  return `${out}"`;
}
